### PR TITLE
[Code Quality] Fixes #4286 Skip ArrayKeyExistsTernaryThenValueToCoalescingRector when else is a not null

### LINKED ISSUE #4286: ArrayKeyExistsTernaryThenValueToCoalescingRector misbehaves
# Bug Report

| Subject        | Details                                                         |
| :------------- | :---------------------------------------------------------------|
| Rector version | e.g. 0.8.7            |
| Installed as   | composer dependency                       |

`ArrayKeyExistsTernaryThenValueToCoalescingRector` is applied where it should not.
The rectoreed code is not the same as before.

## Minimal PHP Code Causing Issue

https://getrector.org/demo/a047e490-98e3-4796-bcbf-d6461a480f4b

## Expected Behaviour

`DemoFile->run()` should return `null` instead of `['value' => null]`.

### PR DESCRIPTION
**Pull request by samsonasik**

### CODE REVIEW

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba it is ready for review.

File: rules/code-quality/src/Rector/Ternary/ArrayKeyExistsTernaryThenValueToCoalescingRector.php [hunk 1]
```diff
@@ -76,6 +77,10 @@ public function refactor(Node $node): ?Node
             return null;
         }
 
+        if ($node->else instanceof Variable) {
+            return null;
+        }
```
Comment (TomasVotruba) [inline rules/code-quality/src/Rector/Ternary/ArrayKeyExistsTernaryThenValueToCoalescingRector.php:82]: I think that if `$node->else` is anything but null, it should be skipped.

Could you add also test fixture where else is string or false?

Comment (samsonasik) [reply rules/code-quality/src/Rector/Ternary/ArrayKeyExistsTernaryThenValueToCoalescingRector.php:82]: I will try.

Comment (TomasVotruba) [reply rules/code-quality/src/Rector/Ternary/ArrayKeyExistsTernaryThenValueToCoalescingRector.php:82]: Ah. Seeing this, the else can be anything and it's correct.
I got confused, sorry.


What exactly is the bug here?

**Comment by TomasVotruba** [comment]
Ref #4286 (the id in title is not clickable on phone)

File: rules/code-quality/src/Rector/Ternary/ArrayKeyExistsTernaryThenValueToCoalescingRector.php (hunk 1 above)
Comment (samsonasik) [reply rules/code-quality/src/Rector/Ternary/ArrayKeyExistsTernaryThenValueToCoalescingRector.php:82]: I think when else is a variable, we should skip that, not sure about false or string or other value..., /cc @infabo for verification

Comment (samsonasik) [reply rules/code-quality/src/Rector/Ternary/ArrayKeyExistsTernaryThenValueToCoalescingRector.php:81]: I've added handle else is not null to be skipped.

Comment (samsonasik) [reply rules/code-quality/src/Rector/Ternary/ArrayKeyExistsTernaryThenValueToCoalescingRector.php:81]: The following stil not checked: 

```php
return array_key_exists('value', $v) ? $v['value'] : 'null';
```

I will re-check.

Comment (samsonasik) [reply rules/code-quality/src/Rector/Ternary/ArrayKeyExistsTernaryThenValueToCoalescingRector.php:82]: I've added handling for string "null"

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba it is ready for review.

**Comment by samsonasik** [comment]
The `node->if` need to check null value ?

**Comment by samsonasik** [comment]
I think it is dangerous to assume that `value` never be changed in `$v = ['value' => null];`, for example the `null` value is a dynamic from parameter/db.

**Comment by TomasVotruba** [comment]
> I think it is dangerous to assume that value never be changed in $v = ['value' => null];, for example the null value is a dynamic from parameter/db.

I think you're right. That leads me to a questoin, are there some cases when this rule brings a value?

I can only think of 1 - when the else value is actually a `null`. Only that way the result is the same in reliable way:

https://3v4l.org/JcGTS

File: rules/code-quality/src/Rector/Ternary/ArrayKeyExistsTernaryThenValueToCoalescingRector.php [hunk 2]
```diff
@@ -76,6 +77,10 @@ public function refactor(Node $node): ?Node
             return null;
         }
 
+        if ($this->getValue($node->else) !== 'null' || $node->else instanceof String_) {
```
Comment (TomasVotruba) [inline rules/code-quality/src/Rector/Ternary/ArrayKeyExistsTernaryThenValueToCoalescingRector.php:80]: Btw, you can use helper method `$this->isNull($node->else)` here.

File: rules/code-quality/tests/Rector/Ternary/ArrayKeyExistsTernaryThenValueToCoalescingRector/Fixture/skip_else_string_null.php.inc [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+namespace Rector\CodeQuality\Tests\Rector\Ternary\ArrayKeyExistsTernaryThenValueToCoalescingRector\Fixture;
+
+class SkipElseStringNullValue
+{
+    public function run()
+    {
+        $v = ['value' => null];
+        return array_key_exists('value', $v) ? $v['value'] : 'null';
+    }
```
Comment (TomasVotruba) [inline rules/code-quality/tests/Rector/Ternary/ArrayKeyExistsTernaryThenValueToCoalescingRector/Fixture/skip_else_string_null.php.inc:11]: :+1:

File: rules/code-quality/src/Rector/Ternary/ArrayKeyExistsTernaryThenValueToCoalescingRector.php (hunk 2 above)
Comment (samsonasik) [reply rules/code-quality/src/Rector/Ternary/ArrayKeyExistsTernaryThenValueToCoalescingRector.php:80]: I tried `$this->isNull`, that not work for this use case.

**Comment by samsonasik** [comment]
I've added more fixture for $v["key"] has value and else is null/not

**Comment by samsonasik** [comment]
I've added  fixture for if is null, else is null .

File: rules/code-quality/src/Rector/Ternary/ArrayKeyExistsTernaryThenValueToCoalescingRector.php (hunk 2 above)
Comment (TomasVotruba) [reply rules/code-quality/src/Rector/Ternary/ArrayKeyExistsTernaryThenValueToCoalescingRector.php:80]: You mean like this?

```php
if (! $this->isNull($node->else)) {
    return null;
}
```

Comment (samsonasik) [reply rules/code-quality/src/Rector/Ternary/ArrayKeyExistsTernaryThenValueToCoalescingRector.php]: thank you, implemented.

**Comment by samsonasik** [comment]
I think it is ready.

**Comment by TomasVotruba** [comment]
Agreed, thank you :+1:

**Pull request merged**